src/embeddings/tools/sgns.cpp

```diff
@@ -0,0 +1,573 @@
+
+                    // Holds the target word and surrounding context words.
+                    sgns_window window;
+
+                    // Load target word.
+                    add_next_index(window, buffer.stream_, word_counter, engine, next_random);
+
+                    // Initialize future context words.
+                    while (window.size() < (max_window_size_ + 1))
+                        if (!add_next_index(window, buffer.stream_, word_counter, engine, next_random))
+                            break;
+
+                    // When the window is first loaded, the target word is the
+                    // first element in the window. The next time a word is added
+                    // to the window, this index will increase by one. This will
+                    // continue until this index reaches the middle of the window,
+                    // where it will stay until the algorithm completes.
+                    sgns_window::size_type target_widx = 0;
+
+                    // While there are target words left in the window...
+                    while (!window.empty() && (target_widx < window.size()))
+                    {
+                        // Sample a random window size.
+                        const auto window_size = next_random(engine) % max_window_size_;
+
+                        // Sweep across the window.
+                        for (sgns_window::size_type w = window_size; w < (max_window_size_ * 2 + 1 - window_size); ++w)
+                        {
+                            // Don't process a target word as its own context word.
+                            // Max window size works here because it is window size + 1th index
+                            // and therefore the target word
+                            if (w != max_window_size_)
+                            {
+                                const int64_t context_widx = (int64_t)target_widx - (int64_t)max_window_size_ + (int64_t)w;
+
+                                if (context_widx < 0 || context_widx >= (int64_t)window.size())
+                                    continue;
+
+                                const sgns_net_vector::size_type l1 = window[(sgns_window::size_type)context_widx] * vector_size_;
+
+                                neu1e.assign(vector_size_, 0);
+
+                                // The first sample here (n == 0) will use the real
+                                // target. The rest (n > 0) will be negative samples.
+                               
+                                for (std::size_t n = 0; n < negative_samples_ + 1; n++)
+                                {
+                                    int label = 1;
+                                    sgns_net_vector::size_type target = window[target_widx];
+
+                                    // TODO: Maybe we do the first sample explicity and eliminate the if logic
+                                    // with this loop? Wonder if it is worth trying?
```

Comment:
I actually made this change as part of a refactor to extract some of this core loop into functions. I think it may have marginally helped.